[Sky] 미션4 : 모든 기물 배치하기

aksmf1442 opened the pull request:
안녕하세요. Sky입니다!
제가 봐도 항상 부족한 코드지만 이번에도 리뷰 잘 부탁드립니다!

### 미션4 변경사항

* 미션3에서 Color enum에 representation을 넣어줬었는데, 이번에는 폰만 있는게 아니라 타입별로 표현방식이 따로 있는 것 같아서 PieceType enum을 새로 만들어주고 여기에 representation을 넣어주었습니다. 여기에도 처음에 색깔별로 표현해주고 싶어서 PieceType enum에 컬러별 표현방식을 변수로 표현해주고 메서드로 받을 수 있게 해줬었는데, 이렇게 하니까 클래스간의 결합도가 높아져서 사이드 이펙트가 더 많이 발생하는 느낌이라서 그냥 representaton 변수를 소문자로 하나 두고 Piece 클래스에서 판별할 수 있도록 classifyRepresentation 메서드를 추가하였습니다.

* 팩토리 메서드를 사용해서 인스턴스를 생성하는 메서드들을 만들어주었습니다. 처음에는 createPiece(Color color, PieceType piecetype)과 같이 하나의 메서드를 통해서 모든 기물들을 만들 수 있도록 하였는데, 이렇게 하니까 어떤 기물을 만드는지가 명확하지 않은 것 같아서 타입별 기물 메서드를 만들고 get기물Instance(Color color)처럼 변경하였습니다.

* 기존의 initialize 메서드는 폰만 초기화하고 있었는데 다른 기물들까지 함께 초기화할 수 있도록 메서드를 추가하였습니다.

aksmf1442 committed:
Refactor print method to Board, Add StringUtils for print method of Board

new line문자가 운영체제마다 달라질 수 있어서 기존의 "\n"문자열을 없애고 System.getProperty("line.separator")를 사용
System.getProperty("line.separator")은 따로 utils pakage를 생성하고 이 곳에 StringUtils 클래스를 생성한 후 appendNewLine("문자열")
메소드를 추가하고 Board의 print() 메소드에서 appendNewLine 메서드를 StringUtils에서 import 해와서 기존의 체스판과 동일한 출력 구현

aksmf1442 committed:
Rename Pawn to Piece, Refactor Piece, Add PieceType enum for piece representation

확장성을 위해서 Pawn 클래스를 Piece 클래스로 이름을 변경
기물별 표현방식을 나타내기 위해서 Color enum에서 representation을 제거하고 새롭게 생성한 PieceType enum에서
color별 representation 변수를 만들어줬다.
Piece 클래스의 생성자를 private하게 만들어주고 색과 이름을 받아서 Piece 객체를 생성하는 팩토리 메서드를 구현해줬다.

aksmf1442 committed:
Refactor PieceType enum and Piece class

클래스간의 결합도를 낮추기 위해서 PieceType eunm에서 컬러별 representation 변수를 지우고 represetation 변수 하나만 남겨두고,
getRepresentation의 인자로 Color를 가져와서 분류했었는데 이걸 단순하게 representation만 반환하도록 변경했다.
그리고 이 역할을 Piece 클래스에서 담당할 수 있도록 classifyRepresentation 메서드를 추가했다.

aksmf1442 committed:
Add showBoard method to Board,  Add initializePieces, initializePawns for Board.initailize

전체 기물의 상태를 볼 수 있게 하기 위해서 Board 클래스에 showBoard 메서드를 추가
Board 클래스에 initiallize 메서드를 세분화해서 Pieces, Pawns별로 initiailize 할 수 있게 메서드 추가

aksmf1442 committed:
Refactor createPiece method -> PieceType getInstance method

어떤 기물을 만드는지 명확하게 하기 위해서 createPiece() 메서드를 조금 더 세분화해서 type별 getInstance() 메서드를 새로 만들었다.

aksmf1442 committed:
Simply addPieces

piece들을 각 리스트에 포함시킬 때 중복되는 코드들을 없애기 위해서 addPiece() 메서드를 추가했다.

wheejuni commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -1,76 +1,112 @@
+    private List<Piece> blackPawns = new ArrayList<>();
+    private List<Piece> whitePieces = new ArrayList<>();
+    private List<Piece> blackPieces = new ArrayList<>();
+    private int pieceCount = 0;
```
상기한 4개 `List` 의 크기를 다 더하면 그게 곧 `pieceCount` 아닐까 하는 생각도 드는데요, 어떠세요?

aksmf1442 replied:
아 그렇네요.! 그게 더 깔끔한 것 같아요. 굳이 따로 변수로 빼줄 필요가 없겠네요.. 수정했습니다!

wheejuni commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -1,76 +1,112 @@
-        return whitePawns.get(idx);
+    private void addPiece(List<Piece> pieces, Piece piece) {
+        pieces.add(piece);
+        pieceCount++;
```
`pieceCount` 를 걷어내고 `List` 를 통해 직접 카운트를 관리하면,
이 로직도 필요없어지고, 이 `private` 메소드도 제거할 수 있을 것 같네요.

aksmf1442 replied:
넵. 수정했습니다!

wheejuni commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -1,76 +1,112 @@
+        addBlackPiece(Piece.getRookInstance(Color.BLACK));
+    }
+
+    public String getPiecesResult(List<Piece> pieces) {
```
`public` 은 조금 과한 것 같기도 합니다. 이 메소드가 외부에서 쓰여야 하나요?

aksmf1442 replied:
아 세심한 부분을 신경 못 쓴 것 같네요. 수정했습니다!

wheejuni commented on src/main/java/net/sky/pieces/Piece.java:
```diff
@@ -0,0 +1,73 @@
+
+    private final Color color;
+    private final PieceType pieceType;
+    private final char REPRESENTATION;
```
인스턴스 변수의 네이밍 컨벤션은 **소문자**입니다.
```suggestion
    private final char representation;

aksmf1442 replied:
상수랑 인스턴스 변수랑 혼동을 한 것 같네요. 생각할 수 있게 해주셔서 감사합니다! 계속 변수 네이밍을 이상하게 할 뻔했네요.

wheejuni commented on src/main/java/net/sky/pieces/Piece.java:
```diff
@@ -0,0 +1,73 @@
+        this.REPRESENTATION = classifyRepresentation();
+    }
+
+    public static Piece getPawnInstance(Color color) {
+        return new Piece(color, PieceType.PAWN);
+    }
+
+    public static Piece getKnightInstance(Color color) {
+        return new Piece(color, PieceType.KNIGHT);
+    }
+
+    public static Piece getRookInstance(Color color) {
+        return new Piece(color, PieceType.ROOK);
+    }
+
+    public static Piece getBishopInstance(Color color) {
+        return new Piece(color, PieceType.BISHOP);
+    }
+
+    public static Piece getQueenInstance(Color color) {
+        return new Piece(color, PieceType.QUEEN);
+    }
+
+    public static Piece getKingInstance(Color color) {
+        return new Piece(color, PieceType.KING);
+    }
```
[이 글](https://johngrib.github.io/wiki/factory-method-pattern/) 을 읽어보시고,
지금 블록 지정한 메소드들을 별도의 클래스로 분리하는 설계도 한번 고려해주세요. 
어렵다고 느껴진다면 진행하지 않아도 될 것 같습니다.

aksmf1442 replied:
나름대로 글을 읽고 변경해봤는데 이렇게 하는건지는 잘 모르겠네요. 이상한 부분 있으면 말씀해주시면 감사하겠습니다!

wheejuni commented on src/main/java/net/sky/pieces/Piece.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public boolean isWhite(Piece piece) {
+        if (piece.getColor() == Color.WHITE) {
+            return true;
+        }
+        return false;
```
아래와 같이 간소화할 수 있습니다:
```suggestion
        return piece.getColor() == Color.WHITE;

aksmf1442 replied:
수정했습니다!

wheejuni commented on src/main/java/net/sky/pieces/PieceType.java:
```diff
@@ -0,0 +1,15 @@
+package net.sky.pieces;
+
+public enum PieceType {
+    PAWN('p'), KNIGHT('n'), ROOK('r'), BISHOP('b'), QUEEN('q'), KING('k');
```
인스턴스 변수를 갖는 `enum` 의 경우 한 줄에 하나씩 선언해주세요.
```suggestion
    PAWN('p'),
    KNIGHT('n'),
    ROOK('r'),
    BISHOP('b'),
    QUEEN('q'),
    KING('k');

aksmf1442 replied:
이런 부분 짚어주시는거 너무 감사합니다! 아는게 많이 없다보니까 이런 세심한 피드백이 너무 좋은 것 같아요!

wheejuni commented on src/main/java/net/sky/pieces/PieceType.java:
```diff
@@ -0,0 +1,15 @@
+public enum PieceType {
+    PAWN('p'), KNIGHT('n'), ROOK('r'), BISHOP('b'), QUEEN('q'), KING('k');
+
+    private final char REPRESENTATION;
```
이유는 모르겠지만 여기도 대문자를 사용하셨네요...

aksmf1442 replied:
이것도 수정했습니다!

wheejuni commented on src/main/java/net/sky/utils/StringUtils.java:
```diff
@@ -0,0 +1,13 @@
+
+public class StringUtils {
+
+    public static final String NEWLINE = System.getProperty("line.separator");
```
`System.getProperty()` 가 아무데서나 호출할 수 있는 `public static` 메소드이므로 굳이 이런 추상화는 필요 없어 보이네요.

aksmf1442 replied:
요구사항에 있는 예제를 그대로 하다보니까 이렇게 된 것 같아요. 앞으로는 요구사항에 있는 예제는 참고만하고 최대한 생각하면서 코드를 작성해야겠네요.

wheejuni commented on src/main/java/net/sky/utils/StringUtils.java:
```diff
@@ -0,0 +1,13 @@
+    }
+
+    public static String appendNewLine(String line) {
+        return line + NEWLINE;
```
위 `NEWLINE` 을 제거하고 아래와 같이 작성하는게 더 나아 보이는데 어떠세요?
```suggestion
        return line + System.getProperty("line.separator");

aksmf1442 replied:
넵. 수정했습니다!

wheejuni reviewed (changes requested):
4단계까지 진행하시느라 수고 많으셨습니다. 💯 👍 
큰 설계의 변경은 필요 없어 보이지만, 몇가지 피드백을 남겨 드렸습니다. 
특히 변수명에 대문자를 쓰신 부분은 학습 초기에 잘못된 습관으로 굳어질 수도 있어서, 앞으로도 꼭 주의해 주시면 좋겠네요.

aksmf1442 committed:
Add PieceMaker for piece, Refactor Board, Piece class

color별 piece, pawn의 리스트의 합이 pieceCount의 값과 같으니까 pieceCount변수를 따로 두지 않고 piece, pawn의 리스트를 합친 값을
getPieceCount()로 받을 수 있게 했다.
PieceMaker라는 클래스를 따로 만들어서 기물별 인스턴스를 쉽게 받을 수 있도록 했다.
인스턴스 변수중 일부 네이밍을 대문자로 했었는데 이를 소문자로 변경했다.

aksmf1442 commented:
확실히 리뷰 받다보니까 제가 부족한 부분들이 많이 보이는 것 같네요. 세심한 피드백 감사합니다! 그리고 일단 나름대로 피드백 받은 부분 고쳐봤는데 이상한 곳 있으면 다시 말씀해주시면 고쳐볼게요!

wheejuni commented on src/main/java/net/sky/pieces/PieceMaker.java:
```diff
@@ -0,0 +1,36 @@
+package net.sky.pieces;
+
+public class PieceMaker {
```
👍

wheejuni reviewed (approved):
👍
